xcell/cls/theory.py

```diff
@@ -2,21 +2,54 @@
 import numpy as np
 
 
+class ConcentrationDuffy08M500c(ccl.halos.Concentration):
+    """ Concentration-mass relation by Duffy et al. 2008
+    (arXiv:0804.2486) extended to Delta = 500-critical.
+    Args:
+        mdef (:class:`~pyccl.halos.massdef.MassDef`): a mass
+            definition object that fixes
+            the mass definition used by this c(M)
+            parametrization.
+    """
+    name = 'Duffy08M500c'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationDuffy08M500c, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = ccl.halos.MassDef(500, 'critical')
+
+    def _check_mdef(self, mdef):
+        if (mdef.Delta != 500) or (mdef.rho_type != 'critical'):
+            return True
+        return False
+
+    def _setup(self):
+        self.A = 3.67
+        self.B = -0.0903
+        self.C = -0.51
+
+    def _concentration(self, cosmo, M, a):
+        M_pivot_inv = cosmo.cosmo.params.h * 5E-13
+        return self.A * (M * M_pivot_inv)**self.B * a**(-self.C)
+
+
 class Theory():
     def __init__(self, data):
         self.config = data['cov']['fiducial']
         self.cosmo = None
-        self.get_cosmo_ccl()
-        self.hm_par = self.get_halomodel_params()
+        self.hm_par = None
 
     def get_cosmo_ccl(self):
         if self.cosmo is None:
             self.cosmo = ccl.Cosmology(**(self.config['cosmo']))
         return self.cosmo
 
     def get_halomodel_params(self):
-        if self.cosmo is None:
-            self.get_cosmo_ccl()
+        if self.hm_par is not None:
+            return self.hm_par
+
+        self.get_cosmo_ccl()
```

Comment:
Maybe it's clearer if you substitute `self.cosmo` in this function for `cosmo` and `cosmo = self.get_cosmo_ccl()`. Same in the other functions. We might forget in a few weeks that `get_cosmo_ccl` is also setting `self.cosmo`.

Reply:
It can even make sense to remove `self.cosmo` and use `self._cosmo` to store the cosmology and only access to this by `get_cosmo_ccl`

Reply:
yep, done